client/client.go

```diff
@@ -42,6 +42,8 @@ type Client interface {
 	GetTS(ctx context.Context) (int64, int64, error)
 	// GetTSAsync gets a timestamp from PD, without block the caller.
 	GetTSAsync(ctx context.Context) TSFuture
+	// GetTSAsync gets a timestamp from PD, without block the caller and provided a tsoRequest pool.
+	GetTSAsyncWithPool(ctx context.Context, pool *sync.Pool) TSFuture
```

Comment:
Pool can be used as package scope, no need to pass it every quest, it can save more memory.

Reply:
this just wants to test divide global pool to session pool to avoid lock cost just as the description said

Reply:
but maybe sync.Pool is not very suitable(after a gc will lost pooled request), maybe better add TsoReqAlloc interface back and use previous self-define pool impl https://github.com/pingcap/tidb/pull/11846/commits/43621c2325bf9c8e5d635d73d9c28343d0c5f6f1#diff-928cb614ae092ed4c6f349f4c4c991e4R200 :laughing:

Reply:
If the lock is an issue, we should solved it in client package(may be by using multiple pool) and keep the interface simple.  The Pool implementation is very fast, it has a fast path for get operation internally.

Reply:
@lysu Sounds good to me if it is really necessary.

Reply:
@lysu Could you try try Pool performance of go1.13? They claim to have some improvements.